lib/Conversion/StableHLOToTTIR/StableHLOToTTIR.cpp

```diff
@@ -56,9 +61,125 @@ class StableHLOToTTIROpConversionPattern : public OpConversionPattern<SrcOp> {
   }
 };
 
+template <typename SrcOp, typename Adaptor = typename SrcOp::Adaptor>
+class StableHLOToTTIRReduceOpConversionPattern
+    : public OpConversionPattern<SrcOp> {
+
+  using OpConversionPattern<SrcOp>::OpConversionPattern;
+
+public:
+  LogicalResult
+  matchAndRewrite(SrcOp srcOp, Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    if (!checkBasicLegality(srcOp)) {
+      return failure();
+    }
+
+    mlir::TypeID srcOpTypeID =
+        srcOp.getBody().front().front().getName().getTypeID();
+
+    // Convert based on type.
+    if (srcOpTypeID ==
+        mlir::detail::TypeIDResolver<mlir::stablehlo::AddOp>::resolveTypeID()) {
+      return matchAndRewriteInternal<mlir::tt::ttir::SumOp>(srcOp, adaptor,
+                                                            rewriter);
+    } else if (srcOpTypeID == mlir::detail::TypeIDResolver<
+                                  mlir::stablehlo::MaxOp>::resolveTypeID()) {
+      return matchAndRewriteInternal<mlir::tt::ttir::MaxOp>(srcOp, adaptor,
+                                                            rewriter);
+    }
+
+    return failure();
+  }
+
+private:
+  bool checkBasicLegality(SrcOp &srcOp) const {
+    if (!srcOp.getBody().hasOneBlock()) {
+      // Expecting StableHLO Reduce OP to have one block inside its body.
+      return false;
+    }
+
+    if (srcOp.getBody().front().empty()) {
+      // Expecting StableHLO Reduce OP to have a body operation defined.
+      return false;
+    }
+
+    return true;
+  }
+
+  template <typename DestOp>
+  LogicalResult
+  matchAndRewriteInternal(SrcOp &srcOp, Adaptor &adaptor,
+                          ConversionPatternRewriter &rewriter) const {
+    auto outputType =
+        mlir::cast<RankedTensorType>(srcOp.getResultTypes().front());
+    tensor::EmptyOp outputTensor = rewriter.create<tensor::EmptyOp>(
+        srcOp.getLoc(), outputType.getShape(), outputType.getElementType());
+
+    mlir::ArrayAttr dimArg = rewriter.getArrayAttr(SmallVector<Attribute>(
+        1, rewriter.getI32IntegerAttr(adaptor.getDimensionsAttr()[0])));
+
+    // If someone changes definition of TTIR_ReductionOp this constant will
+    // become outdated, but I currently see no way to get this info (without
+    // manually constructing the adaptor for dest OP).
+    const std::size_t ttirReduceOpOperandsCount = 2;
+    mlir::ArrayAttr operandConstraints =
+        rewriter.getArrayAttr(SmallVector<Attribute>(
+            ttirReduceOpOperandsCount, rewriter.getAttr<OperandConstraintAttr>(
+                                           OperandConstraint::AnyDeviceTile)));
+
+    rewriter.replaceOpWithNewOp<DestOp>(
+        srcOp, outputType, adaptor.getInputs().front(), outputTensor,
+        false /* keep_dim */, dimArg, operandConstraints);
+
+    return success();
+  }
+};
+
 struct ConvertStableHLOToTTIRPass
```

Comment:
Can we move a conversion pass to a separate file? Like we do for TTIRToTTNN, or TTNNToEmitC.